File: app/src/main/java/com/sumian/app/base/BaseFragment.java
package com.sumian.app.base;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.sumian.app.common.helper.ToastHelper;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * created by jzz
 * <p>
 * on 2017/04/01
 * <p>
 * desc: base fragment
 */
public abstract class BaseFragment<Presenter extends BasePresenter> extends Fragment {

    private static final String TAG = "BaseFragment";

    public Bundle mBundle;
    protected View mRootView;
    protected Activity mHostActivity;
    private Unbinder mUnbinder;

    public Presenter mPresenter;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (mHostActivity == null)
            mHostActivity = (Activity) context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
//        setRetainInstance(true);
        mBundle = getArguments();
        initBundle(mBundle);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(getLayoutId(), container, false);
        // Do something
        onBindViewBefore(rootView);
        // Bind view
        this.mUnbinder = ButterKnife.bind(this, rootView);
        // Get savedInstanceState
        if (savedInstanceState != null) {
            onRestartInstance(savedInstanceState);
        }
        // Init
        initWidget(rootView);
        this.mRootView = rootView;
        // Log.e("TAG", "onCreateView: ----->" + mRootView);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initPresenter();
        initData();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        onRelease();
        //this.mUnbinder.unbind();
    }

    protected void onRestartInstance(Bundle savedInstanceState) {

    }

    protected void onBindViewBefore(View rootView) {

    }

    protected void initBundle(Bundle bundle) {

    }

    protected abstract int getLayoutId();

    protected void initWidget(View root) {

    }

    protected void initPresenter() {

    }

    protected void initData() {

    }

    protected void onRelease() {
        if (mPresenter != null) {
            mPresenter.release();
        }
    }

    protected void runOnUiThread(Runnable run) {
        runOnUiThread(run, 0);
    }

    protected void runOnUiThread(Runnable runnable, long delay) {
        View rootView = this.mRootView;
        if (rootView == null) return;
        if (Looper.myLooper() != Looper.getMainLooper()) {
            rootView.postDelayed(runnable, delay);
        } else {
            if (delay > 0) {
                rootView.postDelayed(runnable, delay);
            } else {
                runnable.run();
            }
        }
    }

    protected void showToast(String message) {
        runOnUiThread(() -> ToastHelper.show(message));
    }

    protected void showToast(@StringRes int messageId) {
        showToast(getString(messageId));
    }

    protected void showCenterToast(String message) {
        if (Looper.myLooper() != Looper.getMainLooper()) {
            runOnUiThread(() -> ToastHelper.show(getContext(), message, Gravity.CENTER));
        } else {
            ToastHelper.show(getContext(), message, Gravity.CENTER);
        }
    }

    protected void showCenterToast(@StringRes int messageId) {
        showCenterToast(getString(messageId));
    }

}

File: common/src/main/java/com/sumian/common/container/base/BaseFragment.kt
package com.sumian.common.container.base

import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.Looper
import android.os.MessageQueue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.sumian.common.container.IContainer
import com.sumian.common.container.delegate.ContainerDelegate
import com.sumian.common.container.delegate.IContainerDelegate
import com.sumian.common.container.mvp.BaseShowLoadingView
import com.sumian.common.container.mvp.IPresenter

/**
 * Created by sm
 *
 * on 2018/12/5
 *
 * desc: fragment container
 *
 */
abstract class BaseFragment<Presenter : IPresenter> : Fragment(), IContainer, BaseShowLoadingView {

    private var mIsActivityCreated = false
    private var mIsViewCreated = false
    private var mIsVisibleToUser = true
    private var mRootView: View? = null

    protected var mPresenter: Presenter? = null

    private val containerDelegate: IContainerDelegate by lazy {
        return@lazy ContainerDelegate.install()
    }

    private val mPrepareInitWidgetTask by lazy {
        MessageQueue.IdleHandler {
            initWidget()
            return@IdleHandler false
        }
    }

    private val mPrepareInitDataTask by lazy {
        MessageQueue.IdleHandler {
            initData()
            return@IdleHandler false
        }
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        super.setUserVisibleHint(isVisibleToUser)
        this.mIsVisibleToUser = isVisibleToUser
        if (mIsActivityCreated && isVisibleToUser) {//用户可见状态，执行任务
            prepareDataTask()
        } else {//用户不可见状态，取消任务
            if (mIsViewCreated) return
            onCancel()
        }
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        this.mIsVisibleToUser = !hidden
        if (hidden) {//调用了 hide commit 为不可见状态，取消任务
            onCancel()
        } else {//调用了show commit  该fragment 为可见状态  执行任务
            if (mIsActivityCreated) {
                prepareDataTask()
            }
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        containerDelegate.setup(context = context!!)
        arguments?.let {
            initBundle(it)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (mRootView == null) {
            mRootView = inflater.inflate(getLayoutId(), container, false)
        }
        return mRootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initWidgetBefore()
        //initWidget()
        prepareWidgetTask()
        mIsViewCreated = true
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        this.mIsActivityCreated = true
        if (mIsViewCreated && mIsVisibleToUser) {
            //initData()
            prepareDataTask()
            mIsViewCreated = false
        }
    }

    override fun onStop() {
        onCancel()
        super.onStop()
    }

    override fun onDestroyView() {
        mIsActivityCreated = false
        mIsViewCreated = false
        mIsVisibleToUser = false
        onRelease()
        super.onDestroyView()
    }

    override fun initWidgetBefore() {
    }

    override fun initBundle(bundle: Bundle) {
    }

    override fun initWidget() {
    }

    override fun initData() {
    }

    override fun onRelease() {
        mPresenter?.onRelease()
        containerDelegate.onRelease()
    }

    override fun onCancel() {
        mPresenter?.onCancel()
        containerDelegate.onCancel()
    }

    override fun onShowLoading() {
        containerDelegate.onShowLoading()
    }

    override fun onDismissLoading() {
        containerDelegate.onDismissLoading()
    }

    private fun prepareDataTask() {
        prepareTask(mPrepareInitDataTask)
    }

    private fun prepareWidgetTask() {
        prepareTask(mPrepareInitWidgetTask)
    }

    private fun prepareTask(task: MessageQueue.IdleHandler) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            context?.mainLooper?.queue?.addIdleHandler(task)
        } else {
            Looper.myQueue().addIdleHandler(task)
        }
    }

    private fun prepareTask(block: () -> Unit) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            context?.mainLooper?.queue?.addIdleHandler {
                block.invoke()
                false
            }
        } else {
            Looper.myQueue().addIdleHandler {
                block.invoke()
                false
            }
        }
    }
}